[ravern] iP

ravern opened the pull request:
# Ekud 🤖

> When in doubt on naming things, simply spell it backwards

**Ekud** is a simple chatbot that helps you remember your to-dos, deadlines and events.

## Features

* Create and delete tasks with _optional_ deadlines or start/end dates
* List all the tasks you currently have
* Mark tasks as done and unmark them as well
* Automatically saves tasks to a file and loads when the chatbot is opened

## Usage

1. Ensure a _recent_ version of Java is installed on your system
2. Download the [latest release](https://github.com/ravern/ip/releases) of Ekud from GitHub
3. Run the JAR file using the command `java -jar Ekud.jar`

## Example

```
Hello! I'm Ekud!
What can I do for you?
> list
Here are the tasks in your list:
1. [E] [ ] Birthday celebration (from: 17:00 to: 21:00)
> todo Clear the trash
Got it. I've added this task:
   [T] [ ] Clear the trash
Now you have 2 tasks in the list.
> list 
Here are the tasks in your list:
1. [E] [ ] Birthday celebration (from: 17:00 to: 21:00)
2. [T] [ ] Clear the trash
> mark 2
Nice! I've marked this task as done:
   [T] [X] Clear the trash
> delete 1
Noted. I've removed this task:
   [E] [ ] Birthday celebration (from: 17:00 to: 21:00)
Now you have 1 tasks in the list.
> list
Here are the tasks in your list:
1. [T] [X] Clear the trash
> bye
Bye. Hope to see you again soon!
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ravern committed:
Rename chatbot to Ekud

ravern committed:
Add echo functionality

ravern committed:
Add adding and listing functionality

ravern committed:
Add (un)marking functionality

ravern committed:
Refactor line processing and split tasks

ravern committed:
Refactor command parsing and add different task types

ravern committed:
Refactor to improve Store abstraction

ravern committed:
Refine main file

ravern committed:
Update tests to latest

ravern committed:
Improve and simplify error handling

ravern committed:
Add delete feature

ravern committed:
Refactor program into state machine

ravern committed:
Add ability to save to a file

ravern committed:
Fix option slash to be forward not backward

Add ignore for data files

ravern committed:
Add basic DateTime parsing

ravern committed:
Add DateTime util to handle parsing and formatting

ravern committed:
Refactor with Ui, Storage, TaskList, Parser class

ravern committed:
Merge branch 'master' into add-gradle-support

ravern committed:
Modify build.gradle for renamed app

ravern committed:
Fix to only write mutating commands

ravern committed:
Add tests for Task class

ravern committed:
Merge branch 'add-gradle-support'

ravern committed:
Add DateTime getters and test

ravern committed:
Add test for Parser

ravern committed:
Add Javadoc for command

ravern committed:
Add Javadoc for handler

ravern committed:
Add Javadoc for state

ravern committed:
Add Javadoc for remaining classes

ravern committed:
Add find command

ravern committed:
Update text ui tests

ravern committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/ekud/state/TaskList.java
#	src/main/java/ekud/ui/Ui.java

ravern committed:
add basic gui impl

ravern committed:
add bot and human images

ravern committed:
refine bot and human messages

ravern committed:
remove text-ui-testing

freddychenyouren2 commented on src/main/java/ekud/Program.java:
```diff
@@ -0,0 +1,89 @@
+     * 
+     * @return Whether the program should continue or end execution.
+     */
+    public boolean step(Command command) {
```
Should the name of this method have "can..." or "is..." ? For example: canContinue().

ravern replied:
I don't think so. `step` here processes a command from the user. Naming a method `canContinue` makes it seem like more of a getter, and implies that you're just returning a private property and not actually processing some actions.

For example,

```java
Program program = new Program();
if (!program.canContinue()) {
  System.exit(1);
}
```

Here I would assume that we're simply checking some sort of internal state whether the program should quit, when actually I'm parsing a user command, modifying state, and printing to stdout accordingly all from within that method call.

freddychenyouren2 commented on src/main/java/ekud/command/Parser.java:
```diff
@@ -0,0 +1,157 @@
+            }
+            case "bye": {
+                return new ByeCommand();
+            }
```
Should the indentation of `case` be reduced?

The current coding style is as such:
```
switch (condition) {
    case ABC:
        statements;
        // Fallthrough
    case DEF:
        statements;
        break;
    case XYZ:
        statements;
        break;
    default:
        statements;
        break;
}
```

Perhaps the following style is more aligned with our CS2103T Java Coding Standard Guide?
```
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}
```

ravern replied:
Good catch, thanks!

freddychenyouren2 commented on src/main/java/ekud/handler/Handler.java:
```diff
@@ -0,0 +1,111 @@
+     * @param ui      The user interface.
+     * @return Whether to continue execution of the program main loop.
+     */
+    public boolean handle(State state, Command command, Ui ui) {
```
Similar small issue to the previous method name that returns a boolean.

ravern replied:
Refer to https://github.com/nus-cs2103-AY2324S1/ip/pull/163#discussion_r1318816451

freddychenyouren2 commented on src/main/java/ekud/state/Task.java:
```diff
@@ -0,0 +1,57 @@
+    protected Task(String title, boolean isDone) {
+        this.title = title;
+        this.isDone = isDone;
+    }
```
Should there be a JavaDoc for the `Task` constructor?

ravern replied:
I'm not sure actually. Its visibility is `protected`. But I'll add one anyway, thanks for the catch.

freddychenyouren2 commented on src/main/java/ekud/state/Task.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    public void mark() {
+        isDone = true;
+    }
```
Might a more descriptive method name such as `markAsDone()` be better for scalability purposes? (Suppose there are multiple versions of marking a particular task)

ravern replied:
Good idea, thanks!

freddychenyouren2 commented on src/main/java/ekud/ui/Ui.java:
```diff
@@ -0,0 +1,86 @@
+            return true;
+        }
+        return handler.handle(command);
+    }
```
Here is another boolean method whose name may need to be renamed slightly 😄

ravern replied:
Refer to https://github.com/nus-cs2103-AY2324S1/ip/pull/163#discussion_r1318816451

freddychenyouren2 commented on src/main/java/ekud/util/DateTime.java:
```diff
@@ -0,0 +1,118 @@
+                return new DateTime(LocalDateTime.parse(data, PARSE_DATE_TIME_FORMATTER));
+            }
+        }
+    }
```
Great Job handling multiple possible date and time formats! 👍

freddychenyouren2 reviewed (commented):
LGTM: Just some minor changes required according to our Java Coding Standard. Overall Good Job!

ylyma reviewed (commented):
Good job! Well-structured and comprehensive handling of possible exceptions. No coding standard violations that I can find.

ravern committed:
Fix gradle build issues for uber jar file

ravern committed:
Add assertions for non-empty text

Before this commit, asserting that the text given to the parseComponent and parseCommand functions was done by checking for non-emptiness and throwing an IllegalArgumentException. While this was desirable behavior, I believe assertions are better suited for the task.

This is because this non-empty invariant should be upholded by the programmer who is building Ekud and not the user of the program. Throwing an exception might result in it getting caught by a regular try-catch and eventually being displayed to the user, which is not the purpose of this check.

Therefore, we switch to using assertions here instead of a simple IllegalArgumentException.

ravern committed:
Remove redundant import

ravern committed:
Fix switch case indentation

ravern committed:
Rename mark and unmark methods to be more descriptive

ravern committed:
Add Javadoc to Task constructor

ravern committed:
Merge pull request #1 from ravern/branch-A-Assertions

Add assertions

ravern committed:
Merge pull request #2 from ravern/branch-A-CodeQuality

Improve code quality

ravern committed:
Add parsing of clean command

ravern committed:
Add hashCode-equals contract for tasks, datetime

ravern committed:
Add processing of clean command

ravern committed:
Merge branch 'master' into branch-C-DetectDuplicates

ravern committed:
Merge pull request #3 from ravern/branch-C-DetectDuplicates

Add clean command to deduplicate task list

ravern committed:
Fix issues caught be soc-se-bot

ravern committed:
Add basic documentation

ravern committed:
Add more screenshots and finish user guide

ravern committed:
Fix distribution issues

ravern committed:
Update README
